### PR TITLE
New A star algorithm pr

### PR DESCRIPTION
**Pull request by weicent**
1. rewrite A star algorithm, improved efficiency by searching the path from two side simultaneously (i.e. from start to goal + from goal to start)

2. generate obstacles map randomly, the number of obstacle is adjustable

Find a path:
FindPath
FindPath2

No path to the goal, indicate the boundary confine robot or goal:
NoPath
NoPath2

### CODE REVIEW

**Comment by AtsushiSakai** [comment]
Thank you for PR. Before review, could you please add a test file for your code?

**Comment by weicent** [comment]
> Thank you for PR. Before review, could you please add a test file for your code?

Hi Atsushi, thanks for your reply. may I ask what is 'test file'?  I'm a newbie in github, it will be helpful if you can give me some hint, thanks in advance!

**Comment by sarimmehdi** [comment]
@weicent use this as reference: https://github.com/AtsushiSakai/PythonRobotics/blob/master/tests/test_a_star.py

**Comment by weicent** [comment]
> @weicent use this as reference: https://github.com/AtsushiSakai/PythonRobotics/blob/master/tests/test_a_star.py

Thanks for let me know this, it helps a lot! The test file is added now!

**Comment by weicent** [comment]
Hi Atsushi, i'm sorry for bother you with so much commits.

It seems that i was trapped with the CI test (together with the dwa_pr, this is my first time to create pull request in github), sorry for the inconvenience for you.

Could you please give me some hint for how to fix the issue? the codecov/patch just indicate some lines are not covered in the test and i don't kown what to do with it.

Thank you in advance.

**Comment by AtsushiSakai** [comment]
@sarimmehdi Thank you for your support : ).

@weicent Thank you for your hard work. No problem, you can ignore the code cov error, because you added a new file.
Let's start reviewing DWA PR first. After that I will check this PR. Thanks again.

**Comment by AtsushiSakai** [comment]
@weicent Hi. Before code review, could you please merger master code to this PR branch?

**Comment by weicent** [comment]
@AtsushiSakai Hi, i have made the rebase for the branch, please consider my pr, thanks in advance!

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+    '''node with properties of g, h, coordinate and parent node'''
+
+    def __init__(self, G=0, H=0, coordinate=None, parent=None):
+        self.G = G
+        self.H = H
+        self.F = G + H
+        self.parent = parent
+        self.coordinate = coordinate
+
+    def setg(self, value):
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:25]: I think these setter functions are not needed for understanding algorithm. You can just access these fields.

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        path_goal.append(node.coordinate)
+        node = node.parent
+    path_goal.append(goal_list[0].coordinate)
+    path_org.reverse()
+    path = path_org + path_goal
+    path = np.array(path)
+    return path
+
+
+def main(obstacle_number=1500):
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:245]: This main function is too long and complicated, could you split some functions?

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+    # find obstacles around node, help to draw the borderline
+    boundary: list = []
+    for x in range(coor[0] - 1, coor[0] + 2):
+        for y in range(coor[1] - 1, coor[1] + 2):
+            if [x, y] in obs:
+                boundary.append([x, y])
+    return boundary
+
+
+def border_line(closed, obs):
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:213]: get_border_line is better?

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+    plt.plot(close_origin[:, 0], close_origin[:, 1], 'oy')
+    plt.plot(close_goal[:, 0], close_goal[:, 1], 'og')
+    plt.plot(bound[:, 0], bound[:, 1], 'sk')
+    plt.plot(end[0], end[1], '*b', label='Goal')
+    plt.plot(start[0], start[1], '^b', label='Origin')
+    plt.legend()
+    plt.pause(0.0001)
+
+
+def node_to_coor(node_list):
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:196]: node_to_coordinate is better?

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
+'''
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:1]: Could you please change all ''' to """ ?

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+    path_goal.append(goal_list[0].coordinate)
+    path_org.reverse()
+    path = path_org + path_goal
+    path = np.array(path)
+    return path
+
+
+def main(obstacle_number=1500):
+    print(__file__ + ' start!')
+    topv = [60, 60]  # top right vertex of boundary
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:247]: top_vertex is better?

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+    botv = [0, 0]  # bottom left vertex of boundary
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:248]: bottom_vertex is better?

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+            target_goal = end
+            while True:
+                # searching from start to end
+                origin_open, origin_close = \
+                    find_path(origin_open, origin_close, target_goal, bound)
+
+                # convert node list into coordinate list and array
+                org_cor_list, org_cor_array = node_to_coor(origin_close)
+
+                if origin_open == []:  # no path condition
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:282]: I can be simpler as "not origin_open"

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+                target_origin = min(origin_open, key=lambda x: x.H).coordinate
+
+                # searching from end to start
+                goal_open, goal_close = \
+                    find_path(goal_open, goal_close, target_origin, bound)
+
+                # convert node list into coordinate list and array
+                goa_cor_list, goa_cor_array = node_to_coor(goal_close)
+
+                if goal_open == []:  # no path condition
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:295]: Same upper comment.

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+
+                # update target for searching from start to end
+                target_goal = \
+                    min(goal_open, key=lambda x: x.H).coordinate
+
+                # check if the searching meet each other
+                og_intersect = \
+                    [coor for coor in org_cor_list if coor in goa_cor_list]
+
+                if og_intersect != []:  # a path is find, og_intersect!=[]
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:306]: same upper comment.

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+                        plt.plot(path[:, 0], path[:, 1], '-r')
+                        plt.title('Robot Arrived', size=20, loc='center')
+                    raise Break()
+                if show_animation:
+                    draw(org_cor_array, goa_cor_array, start, end, bound)
+
+        except Break as success:
+            success.print_info()
+    except NoPath as fail:
+        if origin_open == []:
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:321]: same upper comment.

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+            # if origin confined, find border for origin
+            border = border_line(org_cor_list, obst)
+        else:
+            # else goal confined, find border for goal
+            border = border_line(goa_cor_list, obst)
+        if org_cor_list == []:
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:327]: same upper comment.

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+            # in case start point totally confined by obstacles
+            # in this condition, no neighbor node generated at all,
+            # and search ended with close_list=[]
+            org_cor_list.append(start)
+            org_cor_array = np.array(org_cor_list)
+        if goa_cor_list == []:
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:333]: same upper comment.

**Review by AtsushiSakai — CHANGES_REQUESTED** [review]
Thank you for again. I have some comments. PTAL.

**Comment by weicent** [comment]
@AtsushiSakai Thanks for your patient reply! I have modified the code as your suggestion! Please let me know if you have any further comments!

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+    if flag == 0:
+        node_intersect = check_node_coincide(org_closed, goal_closed)
+        if node_intersect:  # a path is find
+            path = get_path(org_closed, goal_closed, node_intersect[0])
+            stop_loop = 1
+            print('Path is find!')
+            if show_animation:  # draw the path
+                plt.plot(path[:, 0], path[:, 1], '-r')
+                plt.title('Robot Arrived', size=20, loc='center')
+                plt.show()
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:273]: Could you please add `plt.pause(0.01)` before `plt.show()`? In my environment, the final red line does not show  without it.

File: PathPlanning/AStar/A_Star_searching_from_two_side.py [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+            plt.show()
+        elif flag == 2:
+            border = get_border_line(goal_closed, obstacle)
+            plt.plot(border[:, 0], border[:, 1], 'xr')
+            plt.title(info, size=14, loc='center')
+            plt.show()
+    return stop_loop
+
+
+def searching_control(start, end, bound, obstacle):
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/A_Star_searching_from_two_side.py:297]: Can you make this function to return x and y position lists of the final path?

**Review by AtsushiSakai — CHANGES_REQUESTED** [review]
Thank you for again. This is the last review. PTAL.

File: PathPlanning/AStar/A_Star_searching_from_two_side.py (hunk 15 above)
Comment (weicent) [reply PathPlanning/AStar/A_Star_searching_from_two_side.py:301]: Thanks for your suggestion! its a good idea to return path coordinate list in searching control. I have modified as you suggested, the main() function will print path list if show_animation == False now.
Thanks again for your patient review!

File: PathPlanning/AStar/A_Star_searching_from_two_side.py (hunk 14 above)
Comment (weicent) [reply PathPlanning/AStar/A_Star_searching_from_two_side.py:275]: Sure!

**Review by AtsushiSakai — APPROVED** [review]
LGTM. Thank you for great PR!!.

**Pull request merged**